File: cses/03_dynamic_programming/2220_counting-numbers/solution.py
import sys
from functools import lru_cache


def count(x):
    # Đếm các số nguyên n với 0 <= n <= x không có hai chữ số liền kề bằng nhau.
    if x < 0:
        return 0
    s = str(x)
    n = len(s)

    # Digit DP: duyệt các chữ số từ trái sang phải theo trạng thái
    # (pos, prev, started, tight). prev = 10 nghĩa là chưa đặt chữ số nào.
    @lru_cache(maxsize=None)
    def dp(pos, prev, started, tight):
        if pos == n:
            # Cơ sở: đặt hết chữ số, số vừa tạo hợp lệ (kể cả số 0 khi toàn bộ
            # là leading zero) nên đếm là 1.
            return 1
        # Khi tight thì chữ số bị chặn trên bởi chữ số tương ứng của x.
        limit = int(s[pos]) if tight else 9
        total = 0
        for d in range(0, limit + 1):
            ntight = tight and (d == limit)
            if not started:
                if d == 0:
                    # Vẫn là leading zero: chưa bắt đầu số, giữ prev = 10.
                    total += dp(pos + 1, 10, False, ntight)
                else:
                    total += dp(pos + 1, d, True, ntight)
            else:
                # Đã bắt đầu số: cấm chữ số trùng chữ số liền trước.
                if d == prev:
                    continue
                total += dp(pos + 1, d, True, ntight)
        return total

    result = dp(0, 10, False, True)
    dp.cache_clear()
    return result


def main():
    data = sys.stdin.buffer.read().split()
    a = int(data[0])
    b = int(data[1])
    # Đáp án cho đoạn [a, b] tính bằng hiệu hai giá trị tiền tố.
    sys.stdout.write(str(count(b) - count(a - 1)) + "\n")


if __name__ == "__main__":
    main()

File: cses/03_dynamic_programming/2220_counting-numbers/solution.cpp
#include <cstring>
#include <iostream>
#include <string>

using namespace std;

// upperBoundDigits: biểu diễn thập phân của cận trên x cho lần count hiện tại.
string upperBoundDigits;
// memo[pos][prev][started][tight]: bảng ghi nhớ của Digit DP. prev = 10 nghĩa là
// chưa đặt chữ số nào (đang trong vùng leading zero), nên cần 11 giá trị.
long long memo[20][11][2][2];

// Digit DP: đếm các số 0..x không có hai chữ số liền kề bằng nhau, duyệt các chữ
// số từ trái sang phải theo trạng thái (pos, prev, started, tight).
long long digitDp(int position, int previousDigit, bool started, bool tight) {
    // Cơ sở: đã đặt hết chữ số, số vừa tạo hợp lệ nên đếm là 1.
    if (position == static_cast<int>(upperBoundDigits.size())) {
        return 1;
    }

    long long& result = memo[position][previousDigit][started][tight];
    if (result != -1) {
        return result;
    }

    result = 0;
    const int upperDigit = upperBoundDigits[position] - '0';
    // Khi tight thì chữ số bị chặn trên bởi chữ số tương ứng của x.
    const int limit = tight ? upperDigit : 9;
    for (int digit = 0; digit <= limit; ++digit) {
        const bool nextTight = tight && digit == upperDigit;
        if (!started && digit == 0) {
            // Vẫn là leading zero: chưa thực sự bắt đầu số, giữ prev = 10.
            result += digitDp(position + 1, 10, false, nextTight);
        } else {
            // Đã bắt đầu số: cấm chữ số trùng chữ số liền trước.
            if (started && digit == previousDigit) {
                continue;
            }
            result += digitDp(position + 1, digit, true, nextTight);
        }
    }
    return result;
}

// countValid(x) = số các số nguyên trong [0, x] thỏa mãn điều kiện.
long long countValid(long long bound) {
    if (bound < 0) {
        return 0;
    }
    upperBoundDigits = to_string(bound);
    memset(memo, -1, sizeof(memo));
    return digitDp(0, 10, false, true);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    long long lowerBound, upperBound;
    cin >> lowerBound >> upperBound;
    // Đáp án cho đoạn [a, b] tính bằng hiệu hai giá trị tiền tố.
    cout << countValid(upperBound) - countValid(lowerBound - 1) << '\n';
    return 0;
}
